source/bus.c

```diff
@@ -0,0 +1,673 @@
+        aws_linked_list_pop_back(&list->listeners);
+        aws_mem_release(list->allocator, listener);
+    }
+    aws_mem_release(list->allocator, list);
+}
+
+/*
+ * AWS_BUS_SYNC implementation
+ */
+struct bus_sync_impl {
+    struct bus_vtable vtable;
+    struct {
+        /* Map of address -> list of listeners */
+        struct aws_hash_table table;
+    } slots;
+};
+
+static void bus_sync_clean_up(struct aws_bus *bus) {
+    struct bus_sync_impl *impl = bus->impl;
+    aws_hash_table_clean_up(&impl->slots.table);
+    aws_mem_release(bus->allocator, impl);
+}
+
+static int bus_sync_send(struct aws_bus *bus, uint64_t address, void *payload, void (*destructor)(void *)) {
+    struct bus_sync_impl *impl = bus->impl;
+    bus_deliver_msg(bus, address, &impl->slots.table, payload);
+    if (destructor) {
+        destructor(payload);
+    }
+    return AWS_OP_SUCCESS;
+}
+
+static int bus_sync_subscribe(struct aws_bus *bus, uint64_t address, aws_bus_listener_fn *callback, void *user_data) {
+    struct bus_sync_impl *impl = bus->impl;
+    return bus_subscribe(bus, address, &impl->slots.table, callback, user_data);
+}
+
+static int bus_sync_unsubscribe(struct aws_bus *bus, uint64_t address, aws_bus_listener_fn *callback, void *user_data) {
+    struct bus_sync_impl *impl = bus->impl;
+    return bus_unsubscribe(bus, address, &impl->slots.table, callback, user_data);
+}
+
+static struct bus_vtable bus_sync_vtable = {
+    .clean_up = bus_sync_clean_up,
+    .send = bus_sync_send,
+    .subscribe = bus_sync_subscribe,
+    .unsubscribe = bus_sync_unsubscribe,
+};
+
+static void bus_sync_init(struct aws_bus *bus, struct aws_bus_options *options) {
+    (void)options;
+
+    struct bus_sync_impl *impl = bus->impl = aws_mem_calloc(bus->allocator, 1, sizeof(struct bus_sync_impl));
+    impl->vtable = bus_sync_vtable;
+
+    if (aws_hash_table_init(
+            &impl->slots.table, bus->allocator, 8, aws_hash_ptr, aws_ptr_eq, NULL, bus_destroy_listener_list)) {
+        goto error;
+    }
+
+    return;
+
+error:
+    aws_mem_release(bus->allocator, impl);
+}
+
+/*
+ * AWS_BUS_ASYNC implementation
+ */
+struct bus_async_impl {
+    struct bus_vtable vtable;
+    struct {
+        /* Map of address -> list of listeners */
+        struct aws_hash_table table;
+    } slots;
+
+    /* Queue of bus_messages to deliver */
+    struct {
+        struct aws_mutex mutex;
+        /* backing memory for the message free list */
+        void *buffer;
+        void *buffer_end; /* 1 past the end of buffer */
+        /* message free list */
+        struct aws_linked_list free;
+        /* message delivery queue */
+        struct aws_linked_list msgs;
+        /* list of pending adds/removes of listeners */
+        struct aws_linked_list subs;
+    } queue;
+
+    /* dispatch thread */
+    struct {
+        struct aws_thread thread;
+        struct aws_condition_variable notify;
+        struct aws_atomic_var started;
```

Comment:
If you don't wait, you can end up in a situation where you can actually:
* init the bus
* send a bunch of messages (all of which will be ignored because running is false)
* destroy the bus

without ever delivering a single message, even if the bus is reliable

Actually, this raises an interesting question: Should call the destructor on the payload if the message will not be sent because the bus is already closed? Probably right?

Reply:
> Actually, this raises an interesting question: Should call the destructor on the payload if the message will not be sent because the bus is already closed? Probably right?

After thinking on this more, this could unleash Zalgo. This would be a case where the user should check the return value if they care, and if it's an error, clean up the payload. If we call the destructor, and they're expecting it to only be called later from another thread, they might be holding a lock or something that will prevent the dtor from succeeding.

I'm open to other ideas here...

Reply:
but you can just set running=true initially, instead of setting it from the thread.
once the thread starts, its a do-while so it's guaranteed tol process those initial messages